tokio/src/sync/mpsc/bounded.rs

```diff
@@ -849,10 +861,68 @@ impl<T> Sender<T> {
     /// }
     /// ```
     pub async fn reserve(&self) -> Result<Permit<'_, T>, SendError<()>> {
-        self.reserve_inner().await?;
+        self.reserve_inner(1).await?;
         Ok(Permit { chan: &self.chan })
     }
 
+    /// Waits for channel capacity. Once capacity to send `n` messages is
+    /// available, it is reserved for the caller.
+    ///
+    /// If the channel is full, the function waits for the number of unreceived
+    /// messages to become `n` less than the channel capacity. Capacity to send `n`
+    /// message is reserved for the caller. A [`PermitIterator`] is returned to track
+    /// the reserved capacity. You can call this [`Iterator`] until it is exhausted to
+    /// get a [`Permit`] and then call [`Permit::send`].
+    ///
+    /// Dropping [`PermitIterator`] without sending all messages releases the capacity back
+    /// to the channel.
+    ///
+    /// [`PermitIterator`]: PermitIterator
+    /// [`Permit`]: Permit
+    /// [`send`]: Permit::send
+    ///
+    /// # Cancel safety
+    ///
+    /// This channel uses a queue to ensure that calls to `send` and `reserve`
+    /// complete in the order they were requested. Cancelling a call to
+    /// `reserve` makes you lose your place in the queue.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = mpsc::channel(2);
+    ///
+    ///     // Reserve capacity
+    ///     let mut permit = tx.reserve_many(2).await.unwrap();
+    ///
+    ///     // Trying to send directly on the `tx` will fail due to no
+    ///     // available capacity.
+    ///     assert!(tx.try_send(123).is_err());
+    ///
+    ///     // Sending with the permit iterator succeeds
+    ///     permit.next().unwrap().send(456);
+    ///     permit.next().unwrap().send(457);
+    ///
+    ///     // The iterator should now be exhausted
+    ///     assert!(permit.next().is_none());
+    ///     
+    ///     // The value sent on the permit is received
+    ///     assert_eq!(rx.recv().await.unwrap(), 456);
+    ///     assert_eq!(rx.recv().await.unwrap(), 457);
+    /// }
+    /// ```
+    pub async fn reserve_many(&self, n: u32) -> Result<PermitIterator<'_, T>, SendError<()>> {
```

Comment:
All other methods on the mpsc channel use usize for numbers related to the number of permits, so this should also be an usize.

Reply:
The `n` field on the `PermitIterator` is u32 because of the `Semaphore` API. Should I keep the u32 type and do the cast in the `reserve_many` fn ? Or also change the `n` field to usize

Reply:
Just casting is a problem, because it might not fit in an u32. It's a problem that has come up before, see #4446.

Reply:
But because the `Semaphore::acquire_many` currently use u32 there is no other way ? Should we wait #4446 to be resolved ?

Reply:
Well, you don't need #4446 to be fully resolved. You just need the internal batch_semaphore to support it with usize, and you don't need to answer the question of "how should the public API for `Semaphore` look now that we have two almost identical functions?".

But you would probably have to implement that yourself, unfortunately, as I don't think the other issue will be solved anytime soon.

Reply:
But it then implies a modification of the `semaphore` and `rwlock` public API if I change the `batch_semaphore` to usize. Or to cast from u32 to usize ?

Reply:
It would overflow only on 16 bit architecture though and as far as I know tokio is not supporting these.